gdpr/anonymizers/model_anonymizers.py

```diff
@@ -0,0 +1,349 @@
+import hashlib
+import random
+import string
+import warnings
+from typing import (
+    TYPE_CHECKING, Any, Dict, ItemsView, Iterator, KeysView, List, Optional, Tuple, Type, Union, ValuesView
+)
+
+from reversion.revisions import _get_options
+
+from django.conf import settings
+from django.contrib.contenttypes.models import ContentType
+from django.core import serializers
+from django.db import transaction
+from django.db.models import Model, QuerySet
+from django.db.models.fields.related_descriptors import ForwardManyToOneDescriptor, ReverseManyToOneDescriptor
+
+from gdpr.anonymizers.base import FieldAnonymizer, ModelAnonymizerMeta, RelationAnonymizer
+from gdpr.fields import Fields
+from gdpr.models import AnonymizedData, LegalReason
+
+
+if TYPE_CHECKING:
+    from gdpr.purposes.default import AbstractPurpose
+
+FieldList = Union[List, Tuple, KeysView[str]]  # List, tuple or return of dict keys() method.
+FieldMatrix = Union[str, Tuple[Any, ...]]
+
+
+class ModelAnonymizerBase(metaclass=ModelAnonymizerMeta):
+    can_anonymize_qs: bool
+    fields: Dict[str, FieldAnonymizer]
+    _base_encryption_key = None
+
+    class IrreversibleAnonymizerException(Exception):
+        pass
+
+    def __init__(self, base_encryption_key: Optional[str] = None):
+        self._base_encryption_key = base_encryption_key
+
+    @property
+    def model(self) -> Type[Model]:
+        return self.Meta.model  # type: ignore
+
+    @property
+    def content_type(self) -> ContentType:
+        """Get model ContentType"""
+        return ContentType.objects.get_for_model(self.model)
+
+    def __getitem__(self, item: str) -> FieldAnonymizer:
+        return self.fields[item]
+
+    def __contains__(self, item: str) -> bool:
+        return item in self.fields.keys()
+
+    def __iter__(self) -> Iterator[str]:
+        for i in self.fields:
+            yield i
+
+    def keys(self) -> KeysView[str]:
+        return self.fields.keys()
+
+    def items(self) -> ItemsView[str, FieldAnonymizer]:
+        return self.fields.items()
+
+    def values(self) -> ValuesView[FieldAnonymizer]:
+        return self.fields.values()
+
+    def get(self, *args, **kwargs) -> Union[FieldAnonymizer, Any]:
+        return self.fields.get(*args, **kwargs)
+
+    def _get_encryption_key(self, obj, field_name: str):
+        """Hash encryption key from `get_encryption_key` and append settings.SECRET_KEY."""
+        return hashlib.sha256(
+            f'{obj.pk}::{self.get_encryption_key(obj)}::{settings.SECRET_KEY}::{field_name}'.encode(
```

Comment:
Nevím jestli je dobrý nápad to vázat na SECRET_KEY, v dokumentaci se zmiňuje že je možné ho rotovat, to ale pak znemožní deanonymizaci, nepletu-li se. Co udělat pro účely této knihovny vlastní klíč?